File: code/0088. Merge Sorted Array/solution2.java
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int[] auxiliary = new int[m + n];
        
        int p1 = 0, p2 = 0, i = 0;
        while (p1 < m && p2 < n) {
            if (nums1[p1] <= nums2[p2]) {
                auxiliary[i] = nums1[p1];
                i++;
                p1++;
            }
            else {
                auxiliary[i] = nums2[p2];
                i++;
                p2++;
            }
        }
        
        while (p1 < m) {
            auxiliary[i] = nums1[p1];
            i++;
            p1++;
        }
        
        while (p2 < n) {
            auxiliary[i] = nums2[p2];
            i++;
            p2++;
        }
        
        // we could not directly reassign nums1 = auxiliary
        for (i = 0; i < m + n; i++) {
            nums1[i] = auxiliary[i];
        }
    }
}

File: code/0010. Regular Expression Matching/solution2.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.length(), n = p.length();
        vector<vector<bool>> dp(n + 1, vector<bool>(m + 1, false));
        
        dp[0][0] = true;
        
        for (int i = 0; i < n; i++) {
            if (p[i] == '*') {
                dp[i + 1][0] = dp[i - 1][0];
            }
        }
        
        for (int i = 0; i < n; i++) {
            if (p[i] == '*') {
                for (int j = 0; j < m; j++) {
                    dp[i + 1][j + 1] = dp[i - 1][j + 1] || (dp[i + 1][j] && (s[j] == p[i - 1] || p[i - 1] == '.'));
                }
            }
            else {
                for (int j = 0; j < m; j++) {
                    dp[i + 1][j + 1] = dp[i][j] && (s[j] == p[i] || p[i] == '.');
                }
            }
        }
        
        return dp[n][m];
    }
};